desktop/core/src/desktop/js/catalog/optimizer/SqlAnalyzer.ts

```diff
@@ -80,79 +84,56 @@ export default class LocalStrategy implements Optimizer {
     });
   }
 
-  fetchTopJoins({ paths, silenceErrors }: PopularityOptions): CancellablePromise<TopJoins> {
-    const path = paths[0].join('.');
-
-    return new CancellablePromise<TopJoins>((resolve, reject, onCancel) => {
-      contextCatalog
-        .getNamespaces({ connector: this.connector, silenceErrors: !silenceErrors })
-        .then(async (result: { namespaces: Namespace[] }) => {
-          if (!result.namespaces.length || !result.namespaces[0].computes.length) {
-            reject('No namespace or compute found');
-            console.warn(result);
-            return;
-          }
-          const entry = await dataCatalog.getEntry({
-            connector: this.connector,
-            path: path,
-            namespace: result.namespaces[0],
-            compute: result.namespaces[0].computes[0]
-          });
-
-          const sourceMetaPromise = entry.getSourceMeta({ silenceErrors });
-
-          onCancel(() => {
-            sourceMetaPromise.cancel();
-          });
-
-          const sourceMeta = await sourceMetaPromise;
-
-          resolve({
-            values: ((<TableSourceMeta>sourceMeta).foreign_keys || []).map(key => ({
-              totalTableCount: 22,
-              totalQueryCount: 3,
-              joinCols: [{ columns: [path + '.' + key.name, key.to] }],
-              tables: [path].concat(key.to.split('.', 2).join('.')),
-              joinType: 'join'
-            }))
-          });
-        })
-        .catch(reject);
-    });
+  fetchTopJoins(options: PopularityOptions): CancellablePromise<TopJoins> {
+    return this.apiStrategy.fetchTopJoins(options);
```

Comment:
Is there a way to keep the Js Join logic all the time (code that was removed) in this strategy too?
(and it gets augmented by apiStrategy.fetchTopJoins)

i.e. optimize.getJoin is missing a lot of JOINs/dialects, js Join is based on FKs of left autocomplete so get a lot of solid ones for free.

Reply:
(same idea for the Risk too)

Reply:
Ah, indeed, my bad. Will join.

Reply:
Updated with merging of local top joins and risks.